File: src/main/java/cn/edu/nwpu/salarymanagementsystem/pojo/exception/DepartmentLevelException.java
package cn.edu.nwpu.salarymanagementsystem.pojo.exception;

import cn.edu.nwpu.salarymanagementsystem.pojo.data.department.MutableDepartment;

/**
 * 当构造{@link MutableDepartment}时层级不为1且没有设定父部门时抛出此异常。
 *
 * @author UnscientificJsZhai
 */
public class DepartmentLevelException extends RuntimeException {

    private final int level;

    public DepartmentLevelException(int level) {
        this.level = level;
    }

    @Override
    public String toString() {
        if (level > 1) {
            return "Level " + level + " Department must have parent Department!";
        } else if (level == 1) {
            return "Level 1 Department must not have parent Department";
        } else {
            return "Illegal Level: " + level;
        }
    }
}

File: src/main/java/cn/edu/nwpu/salarymanagementsystem/pojo/exception/DepartmentLevelException.kt
package cn.edu.nwpu.salarymanagementsystem.pojo.exception

import cn.edu.nwpu.salarymanagementsystem.pojo.data.department.MutableDepartment

/**
 * 部门信息包含等级。设置父部门时，父部门的等级必须是当前部门的等级-1。
 * 如果不符合此要求，在写入数据库之前会抛出此异常。
 */
class DepartmentLevelException(val parent: MutableDepartment, val child: MutableDepartment) :
    Exception() {

    private val parentLevel: Int
        get() = parent.level
    private val childLevel: Int
        get() = child.level

    override fun toString(): String {
        return "Child Department's Level must be Parent Department's Level minus 1.\ncurrent parent:$parentLevel\ncurrent child:$$childLevel"
    }
}